### PR TITLE
Lazy static cache

### PR DESCRIPTION
**Pull request by moxcodes**
## Proposed changes

Generalize that increasingly frequently used technique of a lazy static cache using a function pointer table. I think we have enough of them piling up to justify a general tool.
This PR also immediately replaces those caches in Swsh... utilities which previously had their own implementation of the technique.

I have a draft sitting around for generalizing this to take extra arguments, as are currently used in the similar implementations for filtering and in PR #1456, but I figured I'd get opinions on the simple version to begin with.

### Types of changes:

- [ ] New feature

### Component:

- [ ] Code

### CODE REVIEW

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
There are a number of code issues but I'd first like to discuss what problem this is trying to solve and why I don't think this solves it. My view is that this is trying to make it so we don't need to repeat the lazy cache code several times. I don't think this alone can solve that problem. In short, there's a good reason I have not implemented this: I don't think it's useful. In most cases we only loop over 1 integer list (number of basis functions) and also loop over basis functions and quadrature type. We cannot loop over the enum classes (easily) in a safe way. This is the main reason why I never saw any value in this code. I'd need to see an implementation that can do the following for me to consider it valuable:

- [ ] For any combination of types with values (e.g. integers, enums, and enum classes) is able to handle sparse and dense ranges of them
- [ ] Has a good interface (I think the interface you've presented is fine, I'm just not sure how easy it will be to meet the first item in this list while maintaining it)
- [ ] Efficiently forwards and inlines function calls, while also being split into an `hpp`-`tpp` pairing to avoid horrid compilation overhead that tensor producting types and values causes.
- [ ] If any possible value in an enum class is missed a compiler warning or error is emitted. (If this is not the case then we will have code be harder to maintain)

Now what might be a reasonable route is to find a nice way to wrap the type of cache presented here by some other generic code that allows looping over enum classes in a semi-reasonable way. However, I think without the ability to either loop over the enum class values or have a second framework that would allow that, I don't think this is a good to introduce.

In summary, I'm overall open to a generalized variant of this cache, but this cache alone I don't think adds enough value to merit being in the code base.

**Comment by moxcodes** [comment]
My opinion is that the lazy cache code is difficult to understand to a newcomer, but its functionality is useful. Even the basic version is already repeated twice in the code base, in Spin-weighted spherical harmonic utilities, and at the point of copying and pasting it a third time for matrices in the linear solve, I decided it was time to generalize.

I'm happy to make immediate attempts at the more elaborate version you suggest, but I'm also a bit confused at the statement that this is out-and-out non-useful. Even when one needs to loop over the basis functions and quadrature (as is done in the filtering action), those loops would be preserved, each one calling a version of this class, potentially with extra arguments. This would still save the code duplication of the function pointer tables. Less savings, I grant, but I think any code duplication reduction is valuable.

**Comment by nilsdeppe** [comment]
I don't think it adds clarity when used inside those outer compile-time loops, or would only potentially add clarity to someone who both understands the `LazyCache` class and the loops over enums. This means now someone needs to understand two things instead of one, and how they interact. The reason I see it as an issue is you now have some black box thing with extra template parameters that you need to figure out what they do when you've already established a pattern of looping with functions until you get to a point where you can have a static variable. Basically, one is going from a simple nested function call algorithm to a nested function call algorithm calling into some other class that does more nesting depending on its template parameters, but does not allow you freedom to do so sparsely even though the whole point of caching in this way is to be sparse. 

A note: if `LazyCache` were to replace the existing `StaticCache` I would be okay with leaving it as is for now, but I think that's a large enough change that it should be discussed at a SpECTRE call, not just PR comments. Another move I'd be more okay with is seeing if the code can be combined and a bool or some other flag can be passed at compile time that caching is either lazy or eager.

**Comment by moxcodes** [comment]
I suppose one significant advantage for the general version I saw was having true documentation for the usage and process - previous versions were always in detail or anonymous namespaces, so didn't have a clear place to document.

Can you expand on the point that this version is not sparse in the way that existing versions are sparse? I had thought to implement the multi-dimensional version similarly sparse, only generating the combinations requested, but perhaps I have misunderstood the machinery I constructed.

**Comment by moxcodes** [comment]
Having put a little thought into it, I have three alternatives to propose moving forward with this:
1. Add 'cap' items to the relevant enums and enum classes (Basis, Quadrature, others?), putting a "End" item at the end. This is the only reasonable way I've seen when investigating this to permit the iteration over them in C++.  With those, I think I can put together a reasonable full solution to the design you request. Because the items in enum classes are not actually types (only the full enum class itself), I think there's no language way to accomplish a compile-time check, even with the cap items.

2. Build out an alternative to language enums/enum classes altogether that have all of the desired features, either via some template work or macros. This would be delicate enough that I'd want to discuss on a call precisely what the feature set would be for the enum alternative.

3. Because this actually fully generalizes all use cases that I foresee for SWSH/CCE uses (2 current, and at least one in code waiting to PR), we could consider calling this a SWSH utility and limiting its scope until a good solution for full generality is found.

**Comment by moxcodes** [comment]
@nilsdeppe I've made the change of just making the lazy static cache the inner workings of `StaticCache`. I'm happy with how it shakes out for the swsh utilities.
Just to see how it would work out, I've pulled basis and quadrature listing utilities out of `Projections.cpp` into the public interface and applied tricks from the same for the static cache in `DiscontinuousGalerkin/Filtering.hpp`. Have a look and see how distasteful you find it, and I'll happily either drop the last two commits, or add more applying the same methods to the caches used in the limiters.

File: src/Utilities/StaticCache.hpp [hunk 1]
```diff
@@ -30,7 +30,7 @@
 template <typename Generator, typename T, typename... Ranges>
 class StaticCache {
  public:
-  StaticCache(Generator&& generator) : generator_{generator} {}
```
Comment (nilsdeppe) [inline src/Utilities/StaticCache.hpp]: can you please move these changes to the static cache into the previous commit where you add the cache?

**Review by nilsdeppe — COMMENTED** [review]
I really the first two commits! Nice job! I've thought a bit about how to handle enums generically inside the cache and this is what I've come up with (I like to make sure my ideas work instead of sending people on a crazy TMP adventure that may not have a road to the end...):

```cpp
// Distributed under the MIT License.
// See LICENSE.txt for details.

#pragma once

#include <array>
#include <cstddef>
#include <utility>

#include "ErrorHandling/Assert.hpp"
#include "Utilities/Gsl.hpp"
#include "Utilities/Requires.hpp"
#include "Utilities/TypeTraits.hpp"

/// \ingroup UtilitiesGroup
/// Range of values for StaticCache indices.  The `Start` is inclusive
/// and the `End` is exclusive.  The range must not be empty.
template <size_t Start, size_t End>
struct CacheRange {
  static_assert(Start < End, "CacheRange must include at least one value");
  constexpr static size_t start = Start;
  constexpr static size_t end = End;
  constexpr static size_t size = end - start;
  using value_type = size_t;
};

/// \ingroup UtilitiesGroup
/// Possible enumeration values for the StaticCache. Only values specified here
/// are retrievable.
///
/// \note The `EnumerationType` must be streamable.
template <typename EnumerationType, EnumerationType... Enums>
struct CacheEnumeration {
  constexpr static size_t size = sizeof...(Enums);
  using value_type = EnumerationType;
};

/// \ingroup UtilitiesGroup
/// A cache of objects intended to be stored in a static variable.
///
/// Objects can be accessed using several `size_t` arguments in the
/// ranges specified as template parameters.  The CacheRange template
/// parameters give first and one-past-last values for the valid
/// ranges for each argument.
///
/// \example
/// \snippet Test_StaticCache.cpp static_cache
///
/// \see make_static_cache
///
/// \tparam T type held in the cache
/// \tparam Ranges ranges of valid indices
template <typename Generator, typename T, typename... Ranges>
class StaticCache {
 public:
  template <typename Gen>
  explicit StaticCache(Gen&& generator)
      : generator_{std::forward<Gen>(generator)} {}

  template <typename... Args>
  const T& operator()(const Args... parameters) const noexcept {
    static_assert(sizeof...(parameters) == sizeof...(Ranges),
                  "Number of arguments must match number of ranges.");
    return unwrap_cache(generate_tuple<Ranges>(parameters)...);
  }

 private:
  template <typename Range, typename T1,
            Requires<not std::is_enum<T1>::value> = nullptr>
  auto generate_tuple(const T1 parameter) const noexcept {
    static_assert(
        tt::is_integer_v<std::decay_t<T1>>,
        "The parameter passed for a CacheRange must be an integer type.");
    return std::make_tuple(static_cast<size_t>(parameter),
                           std::integral_constant<size_t, Range::start>{},
                           std::make_index_sequence<Range::size>{});
  }

  template <typename Range, typename T1,
            Requires<std::is_enum<T1>::value> = nullptr>
  std::tuple<std::decay_t<T1>, Range> generate_tuple(const T1 parameter) const
      noexcept {
    static_assert(
        std::is_same<typename Range::value_type, std::decay_t<T1>>::value,
        "Mismatched enum parameter type and cached type.");
    return {parameter, Range{}};
  }

  template <typename... IntegralConstantValues>
  const T& unwrap_cache() const noexcept {
    static const T cached_object = generator_(IntegralConstantValues::value...);
    return cached_object;
  }

  template <typename... IntegralConstantValues, size_t IndexOffset,
            size_t... Is, typename... Args>
  const T& unwrap_cache(
      std::tuple<size_t, std::integral_constant<size_t, IndexOffset>,
                 std::index_sequence<Is...>>
          parameter0,
      Args... parameters) const noexcept {
    if (UNLIKELY(IndexOffset > std::get<0>(parameter0) or
                 std::get<0>(parameter0) >= IndexOffset + sizeof...(Is))) {
      ERROR("Index out of range: " << IndexOffset
                                   << " <= " << std::get<0>(parameter0) << " < "
                                   << IndexOffset + sizeof...(Is));
    }
    // note that the act of assigning to the specified function pointer type
    // fixes the template arguments that need to be inferred.
    static const std::array<
        const T& (StaticCache<Generator, T, Ranges...>::*)(Args...) const,
        sizeof...(Is)>
        cache{{&StaticCache<Generator, T, Ranges...>::unwrap_cache<
            IntegralConstantValues...,
            std::integral_constant<size_t, Is + IndexOffset>>...}};
    // The array `cache` holds pointers to member functions, so we dereference
    // the pointer and invoke it on `this`.
    return (this->*gsl::at(cache, std::get<0>(parameter0) - IndexOffset))(
        parameters...);
  }

  template <typename... IntegralConstantValues, typename EnumType,
            EnumType... EnumValues, typename... Args>
  const T& unwrap_cache(
      std::tuple<EnumType, CacheEnumeration<EnumType, EnumValues...>>
          parameter0,
      Args... parameters) const noexcept {
    size_t array_location = std::numeric_limits<size_t>::max();
    const std::array<EnumType, sizeof...(EnumValues)> values{{EnumValues...}};
    for (size_t i = 0; i < sizeof...(EnumValues); ++i) {
      if (std::get<0>(parameter0) == gsl::at(values, i)) {
        array_location = i;
        break;
      }
    }
    if (UNLIKELY(array_location == std::numeric_limits<size_t>::max())) {
      ERROR("Uncached enumeration value: " << std::get<0>(parameter0));
    }
    // note that the act of assigning to the specified function pointer type
    // fixes the template arguments that need to be inferred.
    static const std::array<
        const T& (StaticCache<Generator, T, Ranges...>::*)(Args...) const,
        sizeof...(EnumValues)>
        cache{{&StaticCache<Generator, T, Ranges...>::unwrap_cache<
            IntegralConstantValues...,
            std::integral_constant<EnumType, EnumValues>>...}};
    // The array `cache` holds pointers to member functions, so we dereference
    // the pointer and invoke it on `this`.
    return (this->*gsl::at(cache, array_location))(parameters...);
  }

  const Generator generator_;
};

/// \ingroup UtilitiesGroup
/// Create a StaticCache, inferring the cached type from the generator.
template <typename... Ranges, typename Generator>
auto make_static_cache(Generator&& generator) noexcept {
  using CachedType = std::decay_t<decltype(
      generator(std::declval<typename Ranges::value_type>()...))>;
  return StaticCache<std::decay_t<Generator>, CachedType, Ranges...>(
      std::forward<Generator>(generator));
}
```
It's mostly what you have, just some overloads added to handle enums. (I made some minor forwarding and decay adjustments that previously led to additional copying and dangling references).

The test I've written for this is:
```cpp
namespace {
enum class Color { Red, Green, Purple };

std::ostream& operator<<(std::ostream& os, Color t) noexcept {
  switch (t) {
    case Color::Red:
      return os << "Red";
    case Color::Green:
      return os << "Green";
    case Color::Purple:
      return os << "Purple";
    default:
      ERROR("Unknown color");
  }
}

enum class Animal { Goldendoodle, Labradoodle, Poodle };

std::ostream& operator<<(std::ostream& os, Animal t) noexcept {
  switch (t) {
    case Animal::Goldendoodle:
      return os << "Goldendoodle";
    case Animal::Labradoodle:
      return os << "Labradoodle";
    case Animal::Poodle:
      return os << "Poodle";
    default:
      ERROR("Unknown Animal");
  }
}
}  // namespace

SPECTRE_TEST_CASE("Unit.Utilities.StaticCache", "[Utilities][Unit]") {
  const auto enum_generator =
      [](const Color& color, const size_t value = 5,
         const Animal animal = Animal::Goldendoodle) noexcept {
    size_t offset_animal = 0;
    switch (animal) {
      case Animal::Goldendoodle:
        offset_animal = 3000;
        break;
      case Animal::Labradoodle:
        offset_animal = 2000;
        break;
      case Animal::Poodle:
        offset_animal = 1000;
        break;
    };

    switch (color) {
      case Color::Red:
        return value + 100 + offset_animal;
      case Color::Green:
        return value + 200 + offset_animal;
      case Color::Purple:
        return value + 300 + offset_animal;
    };
  };
  const auto enum_cache = make_static_cache<
      CacheEnumeration<Color, Color::Red, Color::Green, Color::Purple>>(
      enum_generator);
  CHECK(enum_cache(Color::Red) == 3105);
  CHECK(enum_cache(Color::Green) == 3205);
  CHECK(enum_cache(Color::Purple) == 3305);

  const auto enum_size_t_cache = make_static_cache<
      CacheEnumeration<Color, Color::Red, Color::Green, Color::Purple>,
      CacheRange<3, 5>>(enum_generator);
  CHECK(enum_size_t_cache(Color::Red, 3) == 3103);
  CHECK(enum_size_t_cache(Color::Red, 4) == 3104);
  CHECK(enum_size_t_cache(Color::Green, 3) == 3203);
  CHECK(enum_size_t_cache(Color::Green, 4) == 3204);
  CHECK(enum_size_t_cache(Color::Purple, 3) == 3303);
  CHECK(enum_size_t_cache(Color::Purple, 4) == 3304);

  const auto enum_size_t_enum_cache = make_static_cache<
      CacheEnumeration<Color, Color::Red, Color::Green, Color::Purple>,
      CacheRange<3, 5>,
      CacheEnumeration<Animal, Animal::Goldendoodle, Animal::Labradoodle,
                       Animal::Poodle>>(enum_generator);
  CHECK(enum_size_t_enum_cache(Color::Red, 3, Animal::Goldendoodle) == 3103);
  CHECK(enum_size_t_enum_cache(Color::Red, 4, Animal::Goldendoodle) == 3104);
  CHECK(enum_size_t_enum_cache(Color::Green, 3, Animal::Goldendoodle) == 3203);
  CHECK(enum_size_t_enum_cache(Color::Green, 4, Animal::Goldendoodle) == 3204);
  CHECK(enum_size_t_enum_cache(Color::Purple, 3, Animal::Goldendoodle) == 3303);
  CHECK(enum_size_t_enum_cache(Color::Purple, 4, Animal::Goldendoodle) == 3304);
  CHECK(enum_size_t_enum_cache(Color::Red, 3, Animal::Labradoodle) == 2103);
  CHECK(enum_size_t_enum_cache(Color::Red, 4, Animal::Labradoodle) == 2104);
  CHECK(enum_size_t_enum_cache(Color::Green, 3, Animal::Labradoodle) == 2203);
  CHECK(enum_size_t_enum_cache(Color::Green, 4, Animal::Labradoodle) == 2204);
  CHECK(enum_size_t_enum_cache(Color::Purple, 3, Animal::Labradoodle) == 2303);
  CHECK(enum_size_t_enum_cache(Color::Purple, 4, Animal::Labradoodle) == 2304);
  CHECK(enum_size_t_enum_cache(Color::Red, 3, Animal::Poodle) == 1103);
  CHECK(enum_size_t_enum_cache(Color::Red, 4, Animal::Poodle) == 1104);
  CHECK(enum_size_t_enum_cache(Color::Green, 3, Animal::Poodle) == 1203);
  CHECK(enum_size_t_enum_cache(Color::Green, 4, Animal::Poodle) == 1204);
  CHECK(enum_size_t_enum_cache(Color::Purple, 3, Animal::Poodle) == 1303);
  CHECK(enum_size_t_enum_cache(Color::Purple, 4, Animal::Poodle) == 1304);
}
```

What do you think?

**Comment by moxcodes** [comment]
@nilsdeppe I think I like the scheme of just specifying every enum member that should be cached. At very least, it seems to better preserve the use-case of c++ enums. I'll merge your code snippet into this PR and use it as the foundation for the enum-based caches.
Cheers!

**Comment by nilsdeppe** [comment]
@moxcodes okay, great! I'm glad you like it!

**Comment by moxcodes** [comment]
@nilsdeppe I've merged your suggestions into the static cache and applied the interface to the filtering cache. Having reviewed the inner workings of the caches for the limiters, I decided to let @fmahebert change those when he thinks it's worth it; they have some nontrivial indexing that I didn't want to risk altering.

File: src/NumericalAlgorithms/Spectral/SwshCoefficients.cpp [hunk 2]
```diff
@@ -28,39 +29,14 @@ CoefficientsMetadata::CoefficientsMetadata(const size_t l_max) noexcept
...
-
-}  // namespace
-
 const CoefficientsMetadata& cached_coefficients_metadata(
     const size_t l_max) noexcept {
-  return precomputed_static_coefficients_impl(
-      l_max, std::make_index_sequence<detail::coefficients_maximum_l_max>{});
+  const static auto lazy_coefficients_cache =
+      make_static_cache<CacheRange<0, detail::coefficients_maximum_l_max>>(
+          [](const size_t generator_l_max) {
```
Comment (nilsdeppe) [inline src/NumericalAlgorithms/Spectral/SwshCoefficients.cpp]: This can be `noexcept`

File: src/Evolution/DiscontinuousGalerkin/Filtering.hpp [hunk 3]
```diff
@@ -243,69 +244,22 @@ class ExponentialFilter<FilterIndex, tmpl::list<TagsToFilter...>> {
...
-                std::make_index_sequence<Spectral::maximum_number_of_points<
-                    Spectral::Basis::Chebyshev>>{});
-          default:
-            ERROR("Missing quadrature in exponential filter matrix action.");
-        }
-      default:
-        ERROR("Missing basis in exponential filter matrix action.");
+    const auto cache_function = [this](const size_t extents,
+                                       const Spectral::Basis basis,
+                                       const Spectral::Quadrature quadrature) {
```
Comment (nilsdeppe) [inline src/Evolution/DiscontinuousGalerkin/Filtering.hpp]: this can be `noexcept`

**Review by nilsdeppe — COMMENTED** [review]
Looks good, couple noexcepts that could be added is all I have. Would it be easy to change the `NumericalAlgorithms/Spectra/Projection.cpp` code to use the enum cache as well?

**Comment by moxcodes** [comment]
lambda `noexcept`s squashed in and I've added a commit switching `Projection.cpp` over to the new interface.

File: src/Utilities/StaticCache.hpp [hunk 4]
```diff
@@ -86,45 +50,116 @@ class StaticCacheImpl<T> {
 ///
```
Comment (nilsdeppe) [inline src/Utilities/StaticCache.hpp:66]: I missed this before, sorry, but I think the dox for `StaticCache` should describe and have an example for the enum and mixed enum-range caching.

**Review by nilsdeppe — COMMENTED** [review]
Sorry, I missed that the dox should be updated. Everything else looks good to me. Thank you for doing this :) I think it's a big improvement to the caching code :+1:

**Comment by moxcodes** [comment]
dox for the enum version added in a fixup.
Cheers!

File: src/Utilities/StaticCache.hpp [hunk 5]
```diff
@@ -38,14 +38,26 @@ struct CacheEnumeration {
 /// \ingroup UtilitiesGroup
 /// A cache of objects intended to be stored in a static variable.
 ///
-/// Objects can be accessed using several `size_t` arguments in the
-/// ranges specified as template parameters.  The CacheRange template
-/// parameters give first and one-past-last values for the valid
-/// ranges for each argument.
+/// Objects can be accessed a combination of several `size_t` arguments in the
```
Comment (nilsdeppe) [inline src/Utilities/StaticCache.hpp]: I think you meant `can be accessed VIA a` (missed `via`)

**Review by nilsdeppe — COMMENTED** [review]
Looks good, I noticed a missing word/grammar thing in the dox but I'm happy for you to squash this and the typo fix :)

**Comment by moxcodes** [comment]
dox fixed and squashed
Cheers!

**Review by nilsdeppe — DISMISSED** [review]

**Review by nilsdeppe — DISMISSED** [review]

File: src/Utilities/StaticCache.hpp [hunk 6]
```diff
@@ -12,119 +12,166 @@
...
-/// ranges specified as template parameters.  The CacheRange template
-/// parameters give first and one-past-last values for the valid
-/// ranges for each argument.
+/// Objects can be accessed via a combination of several `size_t` arguments in
+/// the ranges specified as template parameters of type `CacheRange<start, end>`
+/// and/or enumeration members in the collections of type
+/// `CacheEnumeration<EnumerationType, Members...>`.  The CacheRange template
+/// parameters give first and one-past-last values for the valid ranges for each
+/// argument. The `CacheEnumeration` template parameters are the enumeration
+/// type and an explicit set of every enum member to be cached.
```
Comment (fmahebert) [inline src/Utilities/StaticCache.hpp]: Personally I would find this easier to read if rephrased a bit. Here I've taken a stab at it, trying to use shorter sentences and clearer separation of the size_t vs. enum cases. Feel free to use this, or a different rephrasing...

Objects can be accessed via a combination of `size_t` and `enum` arguments.
The range of each `size_t` argument is specified via a template parameter of type
`CacheRange<start, end>`, giving the first and one-past-last values for the range.
Each `enum` argument is specified by a template parameter of type
`CacheEnumeration<EnumerationType, Members...>`, giving the enumeration
type and an explicit set of every enum member to be cached.

File: src/Utilities/StaticCache.hpp [hunk 7]
```diff
@@ -12,119 +12,166 @@
...
-  /// Initialize the cache.  All objects will be created by calling
-  /// `generator` before the constructor returns.
-  // clang-tidy: misc-forwarding-reference-overload - fixed with
-  // Requires, but clang-tidy can't recognize that.
-  template <typename Generator,
-            Requires<not cpp17::is_same_v<std::decay_t<Generator>,
-                                          StaticCache>> = nullptr>
-  explicit constexpr StaticCache(Generator&& generator) noexcept  // NOLINT
-      : data_(generator) {}
+  template <typename Gen>
```
Comment (fmahebert) [inline src/Utilities/StaticCache.hpp:72]: For my understanding: what is the reason to have the different parameter `Gen` vs. the class' template parameter `Generator`?

File: tests/Unit/Utilities/Test_StaticCache.cpp [hunk 8]
```diff
@@ -27,9 +59,20 @@ SPECTRE_TEST_CASE("Unit.Utilities.StaticCache", "[Utilities][Unit]") {
...
+
+  // explicitly call the cache creation to check its contents
+  cache2(0, 3);
+  cache2(0, 4);
+  cache2(1, 3);
+  CHECK(calls.size() == 3);
+
+  cache2(1, 4);
+  cache2(2, 3);
+  cache2(2, 4);
```
Comment (fmahebert) [inline tests/Unit/Utilities/Test_StaticCache.cpp:79]: Above you've checked this is lazy by checking the generator isn't called until you access the cache.

Could you also check the cache is actually a cache, by calling the cache 2+ times with the same args and checking the generator is only called once?

File: tests/Unit/Utilities/Test_StaticCache.cpp [hunk 9]
```diff
@@ -43,18 +86,89 @@ SPECTRE_TEST_CASE("Unit.Utilities.StaticCache", "[Utilities][Unit]") {
...
+  };
+  /// [static_cache_with_enum]
+  const auto enum_cache = make_static_cache<
+      CacheEnumeration<Color, Color::Red, Color::Green, Color::Purple>>(
+      enum_generator);
+  for (const auto color : {Color::Red, Color::Green, Color::Purple}) {
+    CHECK(enum_cache(color) ==
+          std::make_tuple(3, static_cast<size_t>(color) + 1, 5));
+  }
+  /// [static_cache_with_enum]
```
Comment (fmahebert) [inline tests/Unit/Utilities/Test_StaticCache.cpp]: I think this snippet is actually a little confusing, because the `enum_generator` isn't shown as part of the snippet and is actually quite unintuitive. Would it be reasonable to make a simple `enum_generator` that returns something intuitive (perhaps just the string name of the enum?) and that you can include in the snippet?

File: tests/Unit/Utilities/Test_StaticCache.cpp [hunk 10]
```diff
@@ -43,18 +86,89 @@ SPECTRE_TEST_CASE("Unit.Utilities.StaticCache", "[Utilities][Unit]") {
...
+         {Animal::Goldendoodle, Animal::Labradoodle, Animal::Poodle}) {
+      CHECK(enum_size_t_enum_cache(color, 3, animal) ==
+            std::make_tuple(3 - static_cast<size_t>(animal),
+                            static_cast<size_t>(color) + 1, 3));
+      CHECK(enum_size_t_enum_cache(color, 4, animal) ==
+            std::make_tuple(3 - static_cast<size_t>(animal),
+                            static_cast<size_t>(color) + 1, 4));
+    }
+  }
+  /// [static_cache_with_enum_and_numeric]
```
Comment (fmahebert) [inline tests/Unit/Utilities/Test_StaticCache.cpp]: Same comment here. It's not clear to me if there is a simple way to output from two enums and a size_t, but it's also not clear to me if a test as complicated as this is useful to show as documentation. Perhaps returning a string ("Red3Poodle") and checking against `std::string expected = MakeString{} << color << 3 << animal;` would be easier to understand while still being an interesting test?

File: tests/Unit/Utilities/Test_StaticCache.cpp [hunk 11]
```diff
@@ -43,18 +86,89 @@ SPECTRE_TEST_CASE("Unit.Utilities.StaticCache", "[Utilities][Unit]") {
     ++small_calls;
     return size_t{5};
   });
```
Comment (fmahebert) [inline tests/Unit/Utilities/Test_StaticCache.cpp:94]: I think that right now the documentation doesn't mention that the no-arguments case is also possible... perhaps include a simple case like this one above (but without the counter) as a snippet? Or at least mention it is possible in words.

**Review by fmahebert — COMMENTED** [review]
I have no requests for the substance of the code, but I do have some questions and documentation suggestions.

File: src/Utilities/StaticCache.hpp (hunk 7 above)
Comment (moxcodes) [reply src/Utilities/StaticCache.hpp:72]: it allows a slightly different type structure in the passed argument vs the `decay_t<Generator>` passed as a template argument in the function `make_static_cache`. For instance, some of the current uses of the cache wouldn't type match without the additional flexibility because they pass a const lambda. My understanding is that doing it without the decay_t and template arguments, etc, can cause unnecessary copies of the generator.

Comment (nilsdeppe) [reply src/Utilities/StaticCache.hpp:72]: Even worse, without the decay the member variable can be a reference to an object that goes out of scope and so you get a dangling reference :)

**Comment by moxcodes** [comment]
fixups posted for @fmahebert 's review.
Thanks!

**Comment by fmahebert** [comment]
Fixups LGTM, please squash.

Thanks for doing this! Once this is merged I will update the limiter caches :)

**Review by nilsdeppe — APPROVED** [review]

**Review by fmahebert — APPROVED** [review]
